### PR TITLE
902 - EPPLIB DNS SEC Get and Set DSData and KeyData

### LINKED ISSUE #902: EPP Get and Set Key Data/DNS Data
### Issue Description

Create a Setter and Getter for KeyData/DNSData using the SecDNS extension

- [x] CRUD actions on DNSSEC
  - [x] View existing DS Data or Key Data
  - [x] Add new DS Data or Key Data (but not both)
  - [x] Remove existing DS Data or Key Data
  - [x] Replace existing DS Data or Key Data by indicating which item to remove and which item to add
- [x] Implement tests in test_models_domain's class TestRegistrantDNSSEC

### Additional Context (optional)

Implementation Note: this will need to be parsed when info domain request is sent and added to the cache

### Issue Links

Relates to: #493

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
NOTE - this PR has a current dependency on an open PR on the epplib project.  Once that is resolved and merged, I will pull a new build of epplib into this branch and complete this PR.  This does not need review until that is complete.

## Ticket

Resolves #902 

## Changes

- Getter and setter added for dnssecdata
- dnssecdata added to domain _cache
- test cases written

## Context for reviewers

Getter and setter for DNSSEC in domain.py.  No front end yet exists for managing DNSSEC data, so the only way to test this currently is through command line on sandboxes.  Some sample command line code is below in Setup that can be run in your sandbox.  The setter does not do validation or error checking; this is done at the registry, and will also be done in the forms; so if testing edge cases such as invalid digests, RegistryError will be raised and logged and that is to be expected.

## Setup

Because this PR includes a new build of epplib, you must docker compose build in order for changes to work properly.

To test the getter and setter for epplib, test the getter and setter for the domain attribute dnssecdata.  This can be tested from the command line.  The test_models_domain.py file provides some test data that can be used for testing.  Below is sample code that can be run from a shell on your sandbox:

from epplib.models import common
from epplibwrapper import CLIENT as registry, commands
from registrar.models.domain import Domain

dsData = {
    "keyTag": 1234,
    "alg": 3,
    "digestType": 1,
    "digest": "ec0bdd990b39feead889f0ba613db4adec0bdd99",
}
dnssecExtension = {
    "dsData": [common.DSData(**dsData)],
}
domain, _ = Domain.objects.get_or_create(name="fake.gov")
domain._get_or_create_domain()
domain.dnssecdata = dnssecExtension
print(domain.dnssecdata)

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [x] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [x] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [x] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/models/domain.py [hunk 1]
```diff
@@ -279,6 +280,27 @@ def _create_host(self, host, addrs):
...
+    def dnssecdata(self, _dnssecdata: extensions.DNSSECExtension):
+        updateParams = {
+            "maxSigLife": _dnssecdata.get("maxSigLife", None),
+            "dsData": _dnssecdata.get("dsData", None),
+            "keyData": _dnssecdata.get("keyData", None),
+            "remAllDsKeyData": True,
+        }
+        request = commands.UpdateDomain(name=self.name)
+        extension = commands.UpdateDomainDNSSECExtension(**updateParams)
+        request.add_extension(extension)
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:297]: Is it possible for this to also throw an exception?

File: src/registrar/tests/test_models_domain.py [hunk 2]
```diff
@@ -669,44 +671,357 @@ def test_update_is_unsuccessful(self):
...
-    @skip("not implemented yet")
+        def side_effect(_request, cleaned):
+            return MagicMock(
+                res_data=[self.mockDataInfoDomain],
+                extensions=[
+                    extensions.DNSSECExtension(**self.dnssecExtensionWithDsData)
+                ],
+            )
+
+        patcher = patch("registrar.models.domain.registry.send")
```
Comment (zandercymatics) [inline src/registrar/tests/test_models_domain.py:754]: I think this may potentially be defined already in super().setUp(). This is OK but I'd recommend verifying if there are two patchers running concurrently (like by running .stop() on the old one and seeing if that errs out). I don't think we'd want two of those running concurrently as it could cause other issues

File: src/registrar/tests/test_models_domain.py [hunk 3]
```diff
@@ -669,44 +671,357 @@ def test_update_is_unsuccessful(self):
...
+        def side_effect(_request, cleaned):
+            raise RegistryError(code=ErrorCode.PARAMETER_VALUE_RANGE_ERROR)
+
+        patcher = patch("registrar.models.domain.registry.send")
+        mocked_send = patcher.start()
+        mocked_send.side_effect = side_effect
+
+        # if RegistryError is raised, view formats user-friendly
+        # error message if error is_client_error, is_session_error, or
+        # is_server_error; so test for those conditions
```
Comment (zandercymatics) [inline src/registrar/tests/test_models_domain.py:1017]: (Optional) A nice to have for test case coverage would be to, in either test_views and/or test_admin (where appropriate) add a test that verifies that the UI is returning things correctly. An example would be `test_deletion_ready_fsm_failure` in test_admin.py

Not a necessary change however

**Review by zandercymatics — APPROVED** [review]
LGTM! Just needs a few small code changes but I won't block you here on those. Works well when I tested it through SSH

File: src/epplibwrapper/__init__.py [hunk 4]
```diff
@@ -45,13 +47,15 @@
...
     from epplib.models import common
+    from epplib.responses import extensions
 except ImportError:
     pass
 
 __all__ = [
     "CLIENT",
     "commands",
     "common",
+    "extensions",
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/epplibwrapper/__init__.py:58]: ok, no problem to handle this in merge

File: src/registrar/models/domain.py [hunk 5]
```diff
@@ -975,6 +998,14 @@ def _fetch_cache(self, fetch_hosts=False, fetch_contacts=False):
...
                 cleaned["statuses"] = [status.state for status in cleaned["statuses"]]
 
+            # get extensions info, if there is any
+            # DNSSECExtension is one possible extension, make sure to handle
+            # only DNSSECExtension and not other type extensions
+            returned_extensions = dataResponse.extensions
+            cleaned["dnssecdata"] = None
+            for extension in returned_extensions:
+                if isinstance(extension, extensions.DNSSECExtension):
+                    cleaned["dnssecdata"] = extension
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1008]: There are no plans to handle any other extensions.  But we don't want our code to break in the event of other extensions being added through the cloudflare admin.  So, the code just ignores the other extensions and doesn't update them.

File: src/registrar/tests/test_models_domain.py [hunk 6]
```diff
@@ -669,44 +671,357 @@ def test_update_is_unsuccessful(self):
...
+            "alg": 3,
+            "digestType": 1,
+            "digest": "ec0bdd990b39feead889f0ba613db4adecb4adec",
+        }
+        self.keyDataDict = {
+            "flags": 257,
+            "protocol": 3,
+            "alg": 1,
+            "pubKey": "AQPJ////4Q==",
+        }
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_models_domain.py:714]: good point about possibly needing these in testing views.  i will take a look

File: src/epplibwrapper/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/epplibwrapper/__init__.py:59]: ok, done

File: src/registrar/models/domain.py (hunk 1 above)
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:297]: i am not sure.  wouldn't be a registryerror.  i don't see a need to catch and handle an exception here.

File: src/registrar/tests/test_models_domain.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_models_domain.py:823]: not immediately clear to me where i should be moving this to in common.py.  if there is a need for this to be more available to other tests, it will likely be on my current other ticket, so i will look at possibly moving this on that ticket.

File: src/registrar/tests/test_models_domain.py (hunk 2 above)
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_models_domain.py:866]: ok, good catch.  i am not sure why this has not caused any problems, but i have stopped the other patcher.  it may be that the order of operations was working properly here, but stopping the one patcher in favor of the other is just cleaner, so thank you for this suggestion

File: src/registrar/tests/test_models_domain.py (hunk 3 above)
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_models_domain.py:1141]: since this was done in absence of a UI, this will have to wait, but is a good suggestion
